flagd/pkg/service/flag-evaluation/ofrep/handler_test.go

```diff
@@ -0,0 +1,263 @@
+	}{
+		{
+			name:            "success evaluation",
+			method:          http.MethodPost,
+			input:           bytes.NewReader([]byte{}),
+			mockAnyResponse: []evaluator.AnyValue{successValue},
+			expectedStatus:  http.StatusOK,
+		},
+		{
+			name:            "success & evaluation errors",
+			method:          http.MethodPost,
+			input:           bytes.NewReader([]byte{}),
+			mockAnyResponse: []evaluator.AnyValue{successValue, genericErrorValue, flagNotFoundValue},
+			expectedStatus:  http.StatusOK,
+		},
+		{
+			name:            "only evaluation errors",
+			method:          http.MethodPost,
+			input:           bytes.NewReader([]byte{}),
+			mockAnyResponse: []evaluator.AnyValue{genericErrorValue, flagNotFoundValue},
+			expectedStatus:  http.StatusOK,
+		},
+		{
+			name:            "valid context payload",
+			method:          http.MethodPost,
+			input:           bytes.NewReader([]byte("{\"context\": {}}")),
+			mockAnyResponse: []evaluator.AnyValue{},
+			expectedStatus:  http.StatusOK,
+		},
+		{
+			name:           "invalid context payload",
+			method:         http.MethodPost,
+			input:          bytes.NewReader([]byte("{some invalid context}")),
+			expectedStatus: http.StatusBadRequest,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			eval := mock.NewMockIEvaluator(gomock.NewController(t))
+			if test.mockAnyResponse != nil {
+				eval.EXPECT().
+					ResolveAllValues(gomock.Any(), gomock.Any(), gomock.Any()).
+					Return(test.mockAnyResponse)
+			}
+
+			h := handler{Logger: log, evaluator: eval}
+
+			request, err := http.NewRequest(test.method, "/ofrep/v1/evaluate/flags", test.input)
+			if err != nil {
+				t.Fatalf("error setting up request: %v", err)
+			}
+
+			recorder := httptest.NewRecorder()
+
+			router := mux.NewRouter()
+			router.HandleFunc(bulkEvaluation, h.HandleBulkEvaluation)
+			router.ServeHTTP(recorder, request)
+
+			if test.expectedStatus != recorder.Code {
+				t.Errorf("expected status code %d, but got %d", test.expectedStatus, recorder.Code)
+			}
+		})
+	}
+}
+
+func TestWriteJSONResponse(t *testing.T) {
+	log := logger.NewLogger(nil, false)
+	h := handler{Logger: log}
+
+	tests := []struct {
+		name           string
+		status         int
+		payload        interface{}
+		expectedStatus int
+	}{
+		{
+			name:           "success",
+			status:         http.StatusOK,
+			payload:        successValue,
+			expectedStatus: http.StatusOK,
+		},
+		{
+			name:           "marshaling error",
+			status:         http.StatusOK,
+			payload:        func() {}, // make marshaling to fail
+			expectedStatus: http.StatusInternalServerError,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			recorder := httptest.NewRecorder()
+
+			h.writeJSONToResponse(test.status, test.payload, recorder)
+
+			if test.expectedStatus != recorder.Code {
+				t.Errorf("expected status code %d, but got %d", test.expectedStatus, recorder.Code)
+			}
+
```

Comment:
here we could also add a check if the payload was set by accessing it via

```
payload, err := io.ReadAll(recorder.Body)
```

Reply:
Good suggestion, updated with the commit - https://github.com/open-feature/flagd/pull/1247/commits/83857f55635b1045100706e602e09b8eb4241dd1